### PR TITLE
(closes #1902) Add HoistLocalArraysTrans runtime checks

### LINKED ISSUE #1902: HoistLocalArraysTrans ignores size/bound changes on different calls to the same procedure
In the following excerpt, `kjcnt` changes on different calls to `mpp_init_readbot_strip` causing a runtime error with the current implementation. In addition to checking if the array is allocated, we should also confirm its size needs no change.

```
module mppini
  real(kind=wp), allocatable, dimension(:,:), private :: zbot
  
  contains
  subroutine mpp_init_readbot_strip(kjstr, kjcnt, ldoce)
    integer, intent(in) :: kjstr
    integer, intent(in) :: kjcnt
    logical, dimension(jpiglo,kjcnt), intent(out) :: ldoce
    integer :: inumsave
    integer :: idx
    integer :: idx_1
    integer :: idx_2
    integer :: idx_3
    integer :: idx_4
    integer :: idx_5
    integer :: th_idx
    integer :: nthreads
    ! REAL(wp), DIMENSION(jpiglo,kjcnt) ::   zbot
  
    if (.NOT.ALLOCATED(zbot)) then
      ALLOCATE(zbot(1 : jpiglo, 1 : kjcnt))
    end if
  
    ...
  end subroutine mpp_init_readbot_strip
end module mppini
```

Cheers,
-Nuno

### PR DESCRIPTION
**Pull request by sergisiso**

### CODE REVIEW

File: src/psyclone/psyir/transformations/hoist_local_arrays_trans.py [hunk 1]
```diff
@@ -173,22 +177,60 @@ def apply(self, node, options=None):
...
+            for idx, dim in enumerate(orig_shape):
+                if not isinstance(dim.lower, Literal):
+                    expr = BinaryOperation.create(
+                            BinaryOperation.Operator.NE,
+                            BinaryOperation.create(
+                                BinaryOperation.Operator.LBOUND,
+                                Reference(sym),
+                                Literal(str(idx+1), INTEGER_TYPE)),
+                            dim.lower.copy())
+                    if_expr = BinaryOperation.create(
```
Comment (arporter) [inline src/psyclone/psyir/transformations/hoist_local_arrays_trans.py:201]: This confused me for a while so please could you add a comment saying something like "create() adds the new comparison as a child of the existing 'if not allocated' expression" and again at L213 below.

**Review by arporter — CHANGES_REQUESTED** [review]
Nice fix.
All tests and examples are OK with compilation.
Issue number is not mentioned anywhere.
Ref. guide builds without additional warnings.
No need to update documentation as this is a bug fix.

The only thing I'm requesting is a couple of comments as I was convinced the code didn't work until I looked at the tests and saw that it did!

File: src/psyclone/psyir/transformations/hoist_local_arrays_trans.py (hunk 1 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/hoist_local_arrays_trans.py]: I added the comments and also renamed if_expr to cond_expr as it really is just the conditional statement inside the if.
(I also fixed the doctest for `python -m doctest src/psyclone/psyir/transformations/hoist_local_arrays_trans.py` to pass)

**Comment by sergisiso** [comment]
@arporter This is ready for next review.

**Review by arporter — APPROVED** [review]
Tests and examples are all fine with compilation.
Ref. guide builds with no new warnings.
Will proceed to merge...

**Pull request merged**